File: 112-path-sum/path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if(root==null )
        return false;

        int sum=0;
        boolean ans = fun(root,targetSum,sum);
        return ans;
    }
    public static boolean fun(TreeNode root,int targetSum,int sum)
    {
         if (root == null) return false;
         sum+=root.val;
        if(root.left==null && root.right==null)
        return sum==targetSum;

        
        
        boolean ans1=fun(root.left,targetSum,sum);
        boolean ans2=fun(root.right,targetSum,sum);

        if(ans1 || ans2)
        return true;
         
        return false;

    }
   
}

File: 112-path-sum/path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        if(root==nullptr) return false;
        int sum=0;
        return fun(root,sum,targetSum);

    }
public: 
    bool fun(TreeNode* root, int sum, int targetSum){

        if(root==nullptr) return false;
        sum+=root->val;
        if(sum==targetSum && root->left==nullptr && root->right==nullptr) return true;

        bool left=fun(root->left, sum,targetSum);
        
        bool right=fun(root->right,sum,targetSum);

        if(left || right) return true;
        return false;


    }
};